src/datastore_mad/remotes/lvm.rb

```diff
@@ -0,0 +1,804 @@
+            @xml = xml
+            @xml = REXML::Document.new(@xml).root if @xml.is_a? String
+
+            @ds = lvmds
+
+            super(xml)
+
+            lvname = self.class.lvname(@id)
+
+            if @is_persistent
+                pool = ThinPool.new(@ds.vgname, "#{lvname}-pool")
+                @lv  = ThinLV.new(@ds.vgname, lvname, pool)
+            else
+                @lv = LV.new(@ds.vgname, lvname)
+            end
+        end
+
+        def activate(activate = true)
+            @ds.run_bridge(@lv.activate_sh(activate))
+        end
+
+        # Alias
+        def deactivate
+            activate(false)
+        end
+
+        # Create thin pool + volume. Tries to be atomic, i.e., on failure the half-created LV is
+        # destroyed again.
+        def create(activate = false)
+            @ds.run_bridge(<<~EOF, 'Error creating LV')
+                #{@lv.create_sh(@size, :activate => activate).strip}
+                #{@lv.mkfs_sh(@format, @fs).strip if @fs && @format != 'save_as'}
+            EOF
+        rescue StandardError => e
+            @ds.run_bridge(@lv.delete_sh(:cleanup_pool => true))
+            raise e
+        end
+
+        # Activate the base image and copy
+        def clone_from_path
+            script = <<~EOF
+                sudo lvchange -K -ay #{@path}
+                dd if=/dev/#{@path} of=#{@lv.dev} bs=1M
+                sudo lvchange -an #{@path}
+                #{@lv.activate_sh(false).strip}
+            EOF
+            @ds.run_bridge(script)
+        end
+
+        # Activate persistent image as VM disk in a specific location (host)
+        def activate_and_link(dst)
+            vmdir  = dst.path.dirname
+            script = <<~EOF
+                mkdir -p '#{vmdir}'
+                rm -f '#{vmdir.parent}/.monitor'
+                #{@lv.activate_sh.strip}
+                ln -s '#{@lv.dev}' '#{dst.path}'
+            EOF
+
+            MAD.run(dst.host, lvmsync(script), "Error linking disk #{@lv.lvfname} to #{dst}")
+        end
+
+        # Delete volume (+ snapshots + cleanup thin pool, if it's a persistent image)
+        def delete
+            @ds.run_bridge(@lv.delete_sh(:with_snapshots => true, :cleanup_pool => true),
+                           'Error removing LV')
+        end
+
+    end
+
+    # VM Disk
+    class Disk
+
+        include XmlWrapper
+
+        attr_accessor :host
+        attr_reader :id, :vm, :path, :imageid, :idsid, :sdsid, :size, :is_persistent
+
+        def initialize(xml, vm)
+            @xml = xml
+            @vm  = vm
+
+            @is_persistent = xml_text('PERSISTENT')&.downcase == 'yes'
+
+            @id      = xml_text('DISK_ID', true)
+            @imageid = xml_text('IMAGE_ID')
+            @size    = xml_text('SIZE').to_i
+            @idsid   = xml_text('DATASTORE_ID')
+            @sdsid   = vm.sdsid
+
+            @host = vm.host
+            @path = Pathname.new "/var/lib/one/datastores/#{@sdsid}/#{vm.id}/disk.#{@id}"
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

A cross-site scripting (XSS) vulnerability was detected. User-controlled data is being output without proper sanitization, allowing an attacker to inject malicious scripts. This could lead to session hijacking, malware installation, or phishing attacks. Always sanitize and encode user inputs before including them in the output, using context-appropriate encoding methods. Learn More - https://cwe.mitre.org/data/definitions/79.html.